shared_buffer_pool.go

```diff
@@ -0,0 +1,155 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpc
+
+import "sync"
+
+// SharedBufferPool is a pool of buffers that can be shared, resulting in
+// decreased memory allocation. Currently, in gRPC-go, it is only utilized
+// for parsing incoming messages.
+//
+// # Experimental
+//
+// Notice: This API is EXPERIMENTAL and may be changed or removed in a
+// later release.
+type SharedBufferPool interface {
+	// Get returns a buffer with specified length from the pool.
+	//
+	// The returned byte slice may be not zero initialized.
+	Get(length int) []byte
+
+	// Put returns a buffer to the pool.
+	Put(*[]byte)
+}
```

Comment:
I find it strange and inconvenient that this interface gives you a `[]byte` but requires a `*[]byte` to put it back. This forces the caller to allocate a pointer to put elements back to the pool, which slightly defies the purpose of having a pool.

I think that it should either:
- return a `*[]byte` from `Get(int) *[]byte`: which might be inconvenient because the caller may not have a place where that pointer would be stored, so it would be forced to allocate a new pointer to put the slice back.
- accept a `[]byte` in `Put([]byte)` and let the specific `SharedBufferPool` deal with the situation ([example](https://github.com/prometheus/prometheus/pull/12189)).


There's no need to use pointers to the slice, except because the specific implementation below (just like all of them as today) uses `sync.Pool` which can only hold `interface{}` values. However, we can expect a generic version of `sync.Pool` to come in the next versions of go, which would allow us to solve this implementation detail, hence I wouldn't build the exported contract on that.

So, I'd vote for consistency on `Get(int) []byte` and `Put([]byte)`.

Reply:
@hueypark : What are your thoughts on this?

Reply:
I made commit 1f4bc3550a6ef2915b5c9da55279d73c209575d7 for this.
> accept a []byte in Put([]byte) and let the specific SharedBufferPool deal with the situation (https://github.com/prometheus/prometheus/pull/12189).

Reply:
I've reverted change 1f4bc3550a6ef2915b5c9da55279d73c209575d7 due to the `argument should be pointer-like to avoid allocations (SA6002)`.

Reply:
Currently, I don't perceive any practical benefits, so it seems most sensible to maintain the current status.
In future, should a feature like `generic slice sync.Pool` become available, the `Put([]byte)` could be beneficial.
Perhaps we can consider it after https://github.com/golang/go/issues/47657#issuecomment-1496006853.

Reply:
Fair point. Thanks for the discussion @colega and @hueypark.

Reply:
Note that by reverting that, you're fixing the linter's complaint, but not the underlying issue: you just hid the issue behind and interface call so now the linter can't see it, but it still causes an extra allocation there.

Reply:
Thank you for pointing out. I acknowledge it's not a true fix. We'll need to revisit this for a more effective solution later.